### ISSUE TITLE
Deliver all TXs of a block at once

### ISSUE THREAD

**Issue by thith** [tendermint/tendermint#3745]
Currently, Tendermint deliver tx (deliver_tx) one by one, sequentially.
We are researching an application which attempts to run some of the transactions in the same block in parallel. It would be great if Tendermint offers something like "deliver_all_txs".

**Comment by danil-lashin** [comment]
Hi @thith! 

This was discussed, but there was no decision made. 

https://github.com/tendermint/tendermint/blob/develop/docs/architecture/adr-037-deliver-block.md

**Comment by thith** [comment]
Thanks @danil-lashin, it's great to hear that. Hope that proposal is approved soon.
In our app, each transaction has to declare which state it will access so parallelization is possible.

**Comment by melekes** [comment]
@thith thank you for raising this, however the similar issue https://github.com/tendermint/tendermint/issues/2901 already exist. Feel free to add your thoughts there. Cheers!

Duplicate of https://github.com/tendermint/tendermint/issues/2901

**Comment by ebuchman** [comment]
@thith you can already do this, you just have to buffer the txs in the application so you receive all of them before you start processing them. The responses can be sent back asynchronously, they just have to be sent back in order.

**Comment by thith** [comment]
@ebuchman I see, thank you.
